### ISSUE TITLE
<C-h> (CTRL-H) does not work in the new TUI

### ISSUE THREAD

**Issue by rainerborene** [neovim/neovim#2048]
I use the following mappings on my `.nvimrc` and it does not work using tmux inside urxvt.

``` viml
nnoremap <C-h> <C-w>h " only C-h does not work.
nnoremap <C-j> <C-w>j
nnoremap <C-k> <C-w>k
nnoremap <C-l> <C-w>l
```

env output

```
BROWSER=firefox
COLORFGBG=7;0
COLORTERM=rxvt
DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-zNeYXuznrs,guid=844332d151bf76af0ebfba3754e5f937
DESKTOP_SESSION=i3
DESKTOP_STARTUP_ID=i3/i3-sensible-terminal/725-0-arch_TIME89282
DISPLAY=:0
DOCKER_HOST=unix:///var/run/docker.sock
EDITOR=vim
GDMSESSION=i3
GDM_LANG=en_US.utf8
GPG_TTY=/dev/pts/5
GREP_COLOR=97;45
GTK_MODULES=canberra-gtk-module
HG=/usr/bin/hg
LANG=en_US.utf8
LC_MEASUREMENT=pt_BR.utf8
LC_MONETARY=pt_BR.utf8
LC_NUMERIC=pt_BR.utf8
LC_PAPER=pt_BR.utf8
LC_TIME=pt_BR.utf8
LS_COLORS=no=00;37:fi=01;34:di=00;34:ln=00;36:pi=40;33:so=00;35:do=00;35:bd=40;33;01:cd=40;33;01:or=00;05;37;41:mi=00;05;37;41:su=37;41:sg=30;43:tw=30;42:ow=04;34:st=37;44:ex=00;32:*.cmd=00;33:*.exe=00;33:*.com=00;33:*.btm=00;33:*.bat=00;33:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz=01;31:*.bz2=01;31:*.bzip2=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.apk=01;31:*.gem=01;31:*.jpg=00;35:*.JPG=00;35:*.jpeg=00;35:*.gif=00;35:*.bmp=00;35:*.pbm=00;35:*.pgm=00;35:*.ppm=00;35:*.tga=00;35:*.xbm=00;35:*.xpm=00;35:*.tif=00;35:*.tiff=00;35:*.png=00;35:*.svg=00;35:*.svgz=00;35:*.mng=00;35:*.pcx=00;35:*.dl=00;35:*.xcf=00;35:*.xwd=00;35:*.yuv=00;35:*.cgm=00;35:*.emf=00;35:*.eps=00;35:*.CR2=00;35:*.ico=00;35:*.pdf=00;32:*.ps=00;32:*.txt=00;32:*.html=00;32:*.rst=00;32:*.md=00;32:*.patch=00;32:*.diff=00;32:*.tex=00;32:*.doc=00;32:*.xml=00;32:*.xls=00;32:*.xlsx=00;32:*.doc=00;32:*.docx=00;32:*.ppt=00;32:*.pptx=00;32:*.key=00;32:*.pt=01;32:*.tmpl=01;32:*.in=01;32:*.aac=01;36:*.au=01;36:*.flac=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.m4a=01;36:*.axa=01;36:*.oga=01;36:*.spx=01;36:*.xspf=01;36:*.mov=01;36:*.mpg=01;36:*.mpeg=01;36:*.m2v=01;36:*.mkv=01;36:*.ogm=01;36:*.mp4=01;36:*.m4v=01;36:*.mp4v=01;36:*.vob=01;36:*.qt=01;36:*.nuv=01;36:*.wmv=01;36:*.asf=01;36:*.rm=01;36:*.rmvb=01;36:*.flc=01;36:*.avi=01;36:*.fli=01;36:*.flv=01;36:*.gl=01;36:*.m2ts=01;36:*.divx=01;36:*.webm=01;36:*.axv=01;36:*.anx=01;36:*.ogv=01;36:*.ogx=01;36:*.conf=00;36:*.cnf=00;36:*.cfg=00;36:*.ini=00;36:*.properties=00;36:*.yaml=00;36:*.vcl=00;36:*.c=00;33:*.cpp=00;33:*.py=00;33:*.coffesscript=00;33:*.js=00;33:*.rb=00;33:*.sh=00;33:*.zsh=00;33:*.env=00;33:*.bash=00;33:*.php=00;33:*.java=00;33:*.zcml=00;33:*.db=00;32:*.sql=00;32:*.json=00;32:*.plist=00;32:*.fs=00;32:*.tex=01;37:*.rdf=01;37:*.owl=01;37:*.n3=01;37:*.ttl=01;37:*.nt=01;37:*.torrent=01;37:*.xml=01;37:*Makefile=01;37:*Rakefile=01;37:*build.xml=01;37:*rc=01;37:*.nfo=01;37:*README=01;37:*README.txt=01;37:*readme.txt=01;37:*README.markdown=01;37:*README.md=01;37:*.ini=01;37:*.yml=01;37:*.cfg=01;37:*.conf=01;37:*.c=01;37:*.cpp=01;37:*.cc=01;37:*.log=01;30:*.bak=01;30:*.aux=01;30:*.lof=01;30:*.lol=01;30:*.lot=01;30:*.out=01;30:*.toc=01;30:*.bbl=01;30:*.blg=01;30:*~=01;30:*#=01;30:*.part=01;30:*.incomplete=01;30:*.swp=01;30:*.tmp=01;30:*.temp=01;30:*.o=01;30:*.obj=01;30:*.pyc=01;30:*.pyo=01;30:*.class=01;30:*.cache=01;30:*.egg-info=01;30:
MOZ_PLUGIN_PATH=/usr/lib/mozilla/plugins
SHELL=/usr/bin/fish
SHLVL=4
SSH_AGENT_PID=746
SSH_AUTH_SOCK=/tmp/ssh-J6jf9XAbiTjD/agent.725
TERM=screen-256color
TERMINFO=/usr/share/terminfo
TMUX=/tmp/tmux-1000/default,1161,11
TMUX_PANE=%27
WINDOWID=23068681
WINDOWPATH=1
WINEARCH=win32
XDG_RUNTIME_DIR=/run/user/1000
XDG_SEAT=seat0
XDG_SESSION_DESKTOP=i3
XDG_SESSION_ID=c2
XDG_VTNR=1
__fish_bin_dir=/usr/bin
__fish_datadir=/usr/share/fish
__fish_help_dir=/usr/share/doc/fish
__fish_runtime_dir=/run/user/1000
__fish_sysconfdir=/etc/fish
fish_greeting=
```

**Comment by jwkvam** [comment]
I experience the same on OSX with iterm2.

**Comment by tarruda** [comment]
what is displayed if you type `<c-v><c-h>` in insert mode?

**Comment by jwkvam** [comment]
With nvim, Commit: 0df6b9168e9dc09d3e0ae6050d67b7c31c603782
`<BS>`
With vim 7.4.488
`^H`

**Comment by rainerborene** [comment]
Same here. `<BS>`

**Comment by tarruda** [comment]
any chance you compiled against the wrong libtermkey version? Does `rm -rf .deps build && make && build/bin/nvim` change anything?

**Comment by jwkvam** [comment]
Same behavior for me.
```- downloading...
     src='https://github.com/neovim/libtermkey/archive/8c0cb7108cc63218ea19aa898968eede19e19603.tar.gz'```

**Comment by tarruda** [comment]
what is displayed when you press ctrl+h in the `cat` program?

**Comment by jwkvam** [comment]
`^H`

**Comment by foobacca** [comment]
I use `<C-h>` for `gT` (move to tab left) and just tested and it works as expected.  And `<c-v><c-h>` shows `^H`

I'm using the [PPA version](https://github.com/neovim/neovim/wiki/Installing#ubuntu) of nvim - tried 0.0.0+git201502201750+6~ubuntu14.04.1 and 0.0.0+git201502221955+6~ubuntu14.04.1 - both work fine.

I run in tmux in terminator on Ubuntu.

**Comment by jwkvam** [comment]
FWIW, I bisected the merges and it broke at merge #2039

**Comment by tarruda** [comment]
@jwkvam that PR didn't change anything related to parsing these keys: https://github.com/neovim/neovim/pull/2039/files

As I said, there was a problem with how libtermkey handled ctrl+h or backspace, but it was fixed in a recent commit: https://github.com/neovim/libtermkey/commit/619c0c12828a86dcd388b0915b2e9656d6aac92a

**Comment by geoffharcourt** [comment]
Experiencing the same problem on the latest build (61c98e7e35b18081a8f723406d5ed5f241ddbc96). I'm in iTerm, and the problem happens both inside and outside of tmux. `<c-v><c-h>` prints `<BS>`.

**Comment by geoffharcourt** [comment]
One further detail: this problem happens in nvim with a blank .vimrc.

**Comment by acoffman** [comment]
I'm experiencing exactly this same problem.
It happens on a totally clean install with no plugins:

The entire contents of my `.nvimrc` are
````
" navigate splits
nmap <silent> <C-k> :wincmd k<CR>
nmap <silent> <C-j> :wincmd j<CR>
nmap <silent> <C-h> :wincmd h<CR>
nmap <silent> <C-l> :wincmd l<CR>
````

I'm using OSX 10.10.2 with iTerm2 and NVIM 0.0.0-alpha+201502261923. The problem occurs both inside and outside of tmux and also in Terminal.app. All three other bindings work, but <C-h> does not.

**Comment by acoffman** [comment]
Oh, and `<c-v><c-h>` in insert mode prints `<BS>` in the current buffer.

**Comment by adambiggs** [comment]
Same thing here. Default left-navigation key mapping for the [vim-tmux-navigator](https://github.com/christoomey/vim-tmux-navigator) plugin is broken.

iTerm on OSX 10.10.2. With & without Tmux. Also broken in Terminal.app. Works fine in MacVim.

**Comment by lambdahands** [comment]
Experiencing this as well.

iTerm on OSX 10.10.2

**Comment by af** [comment]
I'm also fighting with trying to bind `<C-h>`. iTerm2 on OS X 10.10.2, using `NVIM-0.0.0-alpha+201503020458`.

It's not a good solution, but if you're looking for a workaround, replacing `<C-h>` with `<bs>` in my mappings did the trick for me.

**Comment by geoffharcourt** [comment]
I think we have enough reports now no one else needs to confirm there's a problem, so no further need to add on to this thread.

**Comment by kopischke** [comment]
@tarruda as everybody concerned seems to be using OS X, [this (somewhat elderly) post](http://hints.macworld.com/article.php?story=20040930002324870) might be of relevance: it suggests that OS X terminals defining a different erase character than Linux in the xterm terminfo can cause issues with Backspace / Ctrl-h (which is the backspace equivalent when you use Emacs readline keybindings).

**Comment by ladislas** [comment]
Another alternative is to use `<C-w-h>` which does the trick for me :)

**Comment by ZyX-I** [comment]
@ladislas What is `w` modifier? Win key?

**Comment by ladislas** [comment]
@ZyX-I my bad, it's `<C-w>h` which means you have to presse the `Ctrl`key + `w` letter key + `h` letter key.

It's just from the remap actually... `nnoremap <C-h> <C-w>h`

**Comment by LuRsT** [comment]
@kopischke, I wasn't going to chime in since there are already a lot of reports for this bug, but I just wanted to add that I'm experiencing the same on Arch linux (Using the AUR package which may be outdated)

**Comment by fishman** [comment]
i'm in archlinux too. this must be because of one the terminal libraries.

in nvim `c-h` produces `<BS>` whereas in vim in produces ``

**Comment by tarruda** [comment]
I'd like to dig the cause for this bug but its hard without reproducing locally. Can someone experiencing the problem paste a log of the keys being sent by the TUI? This will help:

```diff
diff --git a/src/nvim/os/input.c b/src/nvim/os/input.c
index 00efa28..5542218 100644
--- a/src/nvim/os/input.c
+++ b/src/nvim/os/input.c
@@ -142,6 +142,7 @@ bool os_isatty(int fd)

 size_t input_enqueue(String keys)
 {
+  fprintf(stderr, "pressed: %s", keys.data);
   char *ptr = keys.data, *end = ptr + keys.size;

   while (rbuffer_available(input_buffer) >= 6 && ptr < end) {
```

After applying the patch and rebuilding, run `./build/bin/nvim 2> keys.log` , press ctrl+h and then paste keys.log here

**Comment by tarruda** [comment]
I've reproduced the problem in my OSX vm. As the link posted by @kopischke explains, OSX includes wrong terminfo information for xterm-256color(used by iterm), it sets the backspace key description to ctrl+h, which results in libvterm/nvim parsing it as backspace.

My guess is that most programs parse ctrl+h correctly because they ignore the `kbs` description. One way to work around the problem is to fix the terminfo entry for your terminal by replacing `kbs=^H` by `kbs=\177`(ascii del). These commands will create a fixed terminfo entry in your HOME directory:

```sh
infocmp $TERM | sed 's/kbs=^[hH]/kbs=\\177/' > $TERM.ti
tic $TERM.ti
```

**Comment by kopischke** [comment]
@tarruda awesome!

> OSX includes wrong terminfo information for xterm-256color(used by iterm), it sets the backspace key description to ctrl+h, which results in libvterm/nvim parsing it as backspace

Two notes: 

1. `xterm-256-color` is also used by recent versions of Terminal.app (Mavericks and above, IIRC)
2. some Linux distros seem to suffer from the same issue, judging from the comments by @LuRsT and @fishman

**Comment by tarruda** [comment]
> some Linux distros seem to suffer from the same issue, judging from the comments by @LuRsT and @fishman

I imagine thats a good reason to ignore the `kbs` description. While it may break a terminal that has a working terminfo file and uses a different representation for backspace, I doubt anyone will write such a terminal as it would be incompatible with most programs. cc @leonerd

**Comment by adambiggs** [comment]
Thanks @tarruda! Your workaround fixed the issue for me.

Just wondering, is this something you plan on fixing in NeoVim? Or should I add your workaround as a permanent thing to my dotfiles? (or both?)

**Comment by khalidchawtany** [comment]
Great, it works now.

**Comment by tarruda** [comment]
> Just wondering, is this something you plan on fixing in NeoVim? Or should I add your workaround as a permanent thing to my dotfiles? (or both?)

I'm gonna wait for @leonerd's position on this. If he chooses to not ignore `kbs` I will modify the TUI module to parse ctrl+h before handling the data to libtermkey, which will fix the problem without requiring any changes to terminfo.

**Comment by jwhitley** [comment]
I've been experiencing this bug running nvim under tmux on OS X.  I did some investigation, and both the `screen-256color` (under tmux) and `xterm-256color` (under iTerm2)  terminfo entries reproduce this issue.  At least it's consistent!?

**Comment by adambiggs** [comment]
I forgot to test in tmux... @tarruda's workaround fixed it in iTerm, but it's still broken in tmux :(

I'm guessing there's a similar workaround for tmux?

**Comment by kopischke** [comment]
@adambiggs your tmux config probably sets a different `$TERM` than your vanilla terminal. Try repeating the workaround steps while in tmux.

**Comment by fishman** [comment]
i will continue using the following for now until @leonerd chimes in. modifying my terminfo just for c-h everywhere i ssh into vim is kinda harsh. 

     if has('nvim')
         nmap <BS> <C-W>h
     endif

**Comment by leonerd** [comment]
I'm not entirely sure what people are waiting for from me.

Ctrl-H is "ASCII backspace" but terminals ought to be sending ASCII DEL for the Backspace key instead. This may or maynot agree with what their terminfo file says, however, so it's all a huge mess.

If you want it to work, ensure your terminal sends ASCII DEL on the Backspace key, and that this fact agrees with what your terminfo file says. In *practical* terms how you arrange for that to be the case, I really can't say.

**Comment by tarruda** [comment]
@leonerd The problem is that every terminal seems to send ASCII DEL regardless of what terminfo says, and every program(and even the OS line editor) seems ignores the `kbs` terminfo entry also using DEL as backspace. 

> I'm not entirely sure what people are waiting for from me.

We want to know if you will patch libtermkey to also follow this convention of treating DEL as backspace regardless of what is read from terminfo. If not, I will modify neovim TUI to preprocess DEL and ctrl+h before delivering these bytes to libtermkey.

**Comment by caifara** [comment]
@fishman Thanks for that solution, works as a charm (for now)!

**Comment by babygau** [comment]
@lambdahands, @adambiggs 
Just in case you couldn't get it worked with `vim-tmux-navigator`, try this
```bash
if has('nvim')
  nmap <bs> :<c-u>TmuxNavigateLeft<cr>
endif
```
It works for me

**Comment by adambiggs** [comment]
@babygau I found that vim mappings only fix `vim-tmux-navigator` *inside vim*. As soon as you navigate to a Tmux pane that isn't running vim, you can no longer navigate left.

@tarruda's [fix](https://github.com/neovim/neovim/issues/2048#issuecomment-78045837) is the only complete solution I've found so far.

**Comment by badeip** [comment]
I am having similar issues with `:tnoremap <c-h>` on OSX.
This happens in `iTerm`, `Terminal` and (both with- and without) `tmux`, and regardless of what TERM is set to.
pressing `<c-h>` in one of the above mentioned terminal emulators, prints `^H`
pressing `<c-h>` in a regular nvim buffer prints  `<BS>`
pressing `<c-h>` in a `:terminal` buffer prints  `^?`

TERM is set to `xterm-256color`
stty erase is set to `^?`
`infocmp $TERM |grep kbs` prints `kbs=\177`

version: `NVIM 0.0.0-alpha+201504172305 (compiled Apr 18 2015 09:26:25) `

None of the suggested workarounds in thread seems to have any effect.

**Comment by dragn** [comment]
There is a workaround: setting up terminals to send appropriate CSI codes for special combinations.
```
<C-H> : ^[[104;5u
<C-M> : ^[[109;5u
<C-I> : ^[[105;5u
```
For iTerm2 these settings work like charm: http://gyazo.com/132d1ce1fe0e14e62c1a263dde2d374e.
See also: http://www.leonerd.org.uk/hacks/fixterms/
Unfortunately, these settings also break mappings in Vim, as it does not use `libtermkey` :-D

**Comment by justinmk** [comment]
@dragn That is **exactly** the approach we should focus on. Much of the benefit of libtermkey is in the CSI support, let's start using it and documenting the steps needed for users to take advantage of it!

**Comment by badeip** [comment]
@dagn Thanks, this fixes the problem for me.
For future reference, this is how to configure iTerm2 to send the correct CSI for `ctrl+h`
 
**Edit** -> **Preferences** -> **Keys**
Press `+`
Press `Ctrl+h` as **Keyboard Shortcut**
Choose `Send Escape Sequence` as **Action**
Type `[104;5u` for **Esc+**

**Comment by Mudox** [comment]
:+1: you all are life savers!

**Comment by atweiden** [comment]
The  <kbd>Ctrl-h</kbd>, <kbd>Ctrl-j</kbd>, <kbd>Ctrl-k</kbd> and <kbd>Ctrl-l</kbd> keys are only working in nvim outside of tmux. Inside of tmux, none of them work. I'm on Arch Linux with a fairly recent nvim build. I've tried the fix as per https://github.com/neovim/neovim/issues/2048#issuecomment-78045837, but it does not seem to have an effect inside tmux for the <kbd>Ctrl-h</kbd> key.

**Comment by Pyrohh** [comment]
@atweiden besides ctrl-h, everything else works for me in tmux using Arch Linux, so this is presumably related to your terminal. Do you use a custom terminfo entry for your terminal, or perhaps some non-standard thing set via `stty`?

**Comment by atweiden** [comment]
I tried with screen-256color, screen-256color patched with instructions from https://github.com/neovim/neovim/issues/2048#issuecomment-78045837, with my custom [screen-it-256color](https://wiki.archlinux.org/index.php/Tmux#Fix_reverse-video.2Fitalic_mode_in_urxvt) terminfo and with that custom terminfo patched.

@Pyrohh Are you using tmux-navigator by chance? I wonder if it have something to do with that plugin.

**Comment by Pyrohh** [comment]
@atweiden I am not. Have you tried just a raw terminal?

**Comment by atweiden** [comment]
xterm and linux tty

**Comment by badeip** [comment]
@atweiden I am using `tmux-navigator`, and after modifying my terminal according to my previous comment, it works both in nvim and nvim within tmux.
IOW, your problem is most likely due to your terminal not sending the correct CSI.

**Comment by justinmk** [comment]
Speaking of tmux, it can be used to map CSI sequences too! This works:

    bind-key -n C-h send-keys Escape "[104;5u"

I guess the reason this works is that tmux happens to not have the "bug" reported in this issue, so it is able to distinguish ctrl-h from backspace as Vim did.

Side note: The [terminal-overrides](http://stackoverflow.com/a/15218891/152142) feature of tmux useful too (but it solves the inverse case, e.g., it maps input sequences *to* terminal capabilities).

**Comment by dragn** [comment]
I've posted a request on iTerm2 bug tracker to add libtermkey codes to preset: https://gitlab.com/gnachman/iterm2/issues/3519.

**Comment by justinmk** [comment]
@dragn Cool. Note that to distinguish between ctrl-A (i.e., ctrl-shift-a) and ctrl-a, I believe more codes will be needed. Also curious why ctrl-tab and ctrl-shift-tab are missing in your list?

To "edit the plist", you should be able to set up the mappings in iTerm2, then export the plist to xml:

    plutil -convert xml1 ~/Library/Preferences/???.plist -o PresetKeyMappings.plist

**Comment by dragn** [comment]
@justinmk Ctrl-Tab is mapped to switch tabs in iTerm by default (as a global shortcut). Overriding It will disable the global shortcut, which I believe is pretty common in use.
About ctrl-shift- combinations... I'm afraid it's counterproductive to add every possible combinations to Key mappings, the better approach is to add an option to produce correct codes for every possible value.  And if I understand correctly, some work has already been done to parse CSI codes correctly in iTerm itself (see option "Use modern parser" [here](https://iterm2.com/documentation-preferences.html)).
Thanks for the tip about `plutil`, I'll look into it.

**Comment by dragn** [comment]
@justinmk It does interpret it as `Tab` for me also, but I think it's some default mapping of `<C-i>`, because mapping it to something else works.

**Comment by justinmk** [comment]
@dragn In nvim, if you go to insert-mode and type `<c-v><c-i>`, what does it insert? Even if I send the raw sequence via tmux, nvim still interprets it as a tab:

    send-keys Escape "[105;5u"
    send-keys Escape "[73;5u"

**Comment by dragn** [comment]
@justinmk Try using [libtermkey](https://github.com/neovim/libtermkey) sources to debug. Clone the repo, `make` it with `DEBUG=1` and start `demo` program, It will output exactly what terminal is sending and how the library (and `nvim` itself) will interpret the sequence.

**Comment by dragn** [comment]
Here is the PR, for the reference (I'm not sure about the preset name, though...): https://github.com/gnachman/iTerm2/pull/213

**Comment by justinmk** [comment]
@dragn Thanks for that quickstart. `demo` shows this:

```
Key <C-i>
getkey(force=0): buffer 
Driver terminfo yields TERMKEY_RES_NONE
Driver CSI yields TERMKEY_RES_NONE
getkey_simple(force=0) yields TERMKEY_RES_NONE
getkey(force=0): buffer 1b 5b 37 33 3b 35 75 
Driver terminfo yields TERMKEY_RES_NONE
Driver CSI yields TERMKEY_RES_KEY
Unicode codepoint=U+0049 utf8='I' mod=C+00
Key <C-I>
getkey(force=0): buffer 
Driver terminfo yields TERMKEY_RES_NONE
Driver CSI yields TERMKEY_RES_NONE
getkey_simple(force=0) yields TERMKEY_RES_NONE
```

`Key <C-i>` and `Key <C-I>` confirms that `demo` is able to distinguish the sequences (it cannot distinguish `<C-b>` and `<C-B>` unless I configure sequences for them).

Note that I'm using Terminal.app, not iTerm2. I figured out how to force sequences even though they are not available in the GUI preferences dialog (will share this later, it requires a script).

**Comment by atweiden** [comment]
Using libtermkey's demo util has revealed that the Ctrl-hjkl key issue is particular to the system `screen-256color*` terminfo. Using the default `screen-256color` $TERM does, with @tarruda's fix, solve the Ctrl-hjkl issue inside tmux.

However, that fix comes with other issues for vim tmux users.

For instance, libtermkey doesn't detect <kbd>Shift-F[9..12]</kbd> using the `screen-256color` term, whereas with my terminfo [screen-it-256color](https://github.com/atweiden/dotfiles/commit/ec6dd2292289632d3dbebdcb04b7931f78e27ef0), libtermkey can detect those keys. Moreover, `screen-it-256color` is [fixed](https://wiki.archlinux.org/index.php/Tmux#Fix_reverse-video.2Fitalic_mode_in_urxvt) to make Markdown italics render properly in vim on tmux. So for me at least, the Ctrl key bindings are more important, but I also bind <kbd>Shift-F12</kbd> and don't mind italics rendering properly inside tmux.

 Could anyone confirm by running libtermkey's demo util in tmux and pressing <kbd>Shift-F9</kbd>,  <kbd>Shift-F10</kbd>,  <kbd>Shift-F11</kbd> or  <kbd>Shift-F12</kbd>?

**Comment by gnachman** [comment]
Do I understand correctly that in iTerm2 some people get 0x7f when pressing Control and H? That should only be possible if there's a key mapping overriding ^H. This is not to be confused with the Delete key, which will send either 0x7f or 0x08 depending on how it's configured in prefs>profiles>keys.

**Comment by geoffharcourt** [comment]
@gnachmanm, in iTerm2 Build 2.9.20150417-nightly, with no key mappings set in the Keys - Key Mappings section, ^H in insert mode appears to send a backspace. I checked this with no .vimrc loaded.

**Comment by gnachman** [comment]
@geoffharcourt Would you please make a debug log of pressing Control-H? Make sure you use the nightly build for this, as it has good logging. Instructions here:
https://gitlab.com/gnachman/iterm2/wikis/DebugLogging
Send me a link to a gist or drop it as an attachment in this issue:
https://gitlab.com/gnachman/iterm2/issues/3519

**Comment by geoffharcourt** [comment]
Posted a gist there. Thanks.

**Comment by gnachman** [comment]
iTerm2 is sending 0x08 for C-h. I guess the issue that the broken terminfo file interprets this as backspace when 0x7f should be backspace.

**Comment by agauniyal** [comment]
I have same problem regarding backspace key. Also in gnome terminal , which I'm using , there is an option in compatibility which says Backspace key generates ( currently set as ASCII DEL) : 

1. Automatic
2. Control - H
3. ASCII DEL
4. Escape Sequence 
5. TTY Erase

also below that , I have Delete key generates set as escape sequence.

**Comment by justinmk** [comment]
@agauniyal Can you try all 5 gnome-terminal settings and let us know which (if any) fixes CTRL-H in nvim?

**Comment by agauniyal** [comment]
@justinmk , this is weird , I tried out all options and none of them actually fixed backspace key , But whenever I pressed del key , the character that was deleted , could again be deleted by backspace key. I know that sounds confusing so here's a recording : 

https://asciinema.org/a/19874

see from 1:20 to 1:24-5 I was constantly pressing backspace key but no character was being deleted.

**Comment by justinmk** [comment]
@agauniyal for `4. Escape Sequence`, the following should work:

    ^[[104;5u

`^[` is the literal "escape". Make sure you use whatever gnome terminal wants there to represent "escape".

**Comment by agauniyal** [comment]
@justinmk I guess you're asking me to set `4. Escape sequence` as `^[[104;5u` , but gnome terminal dosen't provides a way to do that. It provides a dropdown : 
![2015-05-11-135840_1366x768_scrot](https://cloud.githubusercontent.com/assets/7630575/7560975/e9479728-f7e5-11e4-9ddc-0b0d76ce76b4.png)

Also of I select escape sequence , then backspace key behaves like del key in vim as well as in terminal. But nvim dosent responds to it , completely ignores backspace key till I del a character and press backspace key on it as I have shown above.

**Comment by chevex** [comment]
Why does this just work in regular vim but not in neovim if the issue is with OS X, iTerm, or tmux?

**Comment by agauniyal** [comment]
@chevex  I'm nowhere near OS X , having arch linux with gnome terminal and this dosen't works on mine system too.

**Comment by davidsteinberger** [comment]
I know that this is odd. Did u try this in your vimrc?

    if has('nvim')
      let $NVIM_TUI_ENABLE_CURSOR_SHAPE=1
      " Hack to get C-h working in neovim
      nmap <BS> <C-W>h
      tnoremap <Esc> <C-\><C-n>
    endif

**Comment by chevex** [comment]
@tarruda's fix worked for me great. I'm just curious why vim works and neovim doesn't. That lone seems to say that neovim is what needs to change. Maybe I'm not fully understanding. I'm new to neovim.

**Comment by justinnhli** [comment]
@chevex I believe the issue is that vim and other programs actually ignore what is specified in terminfo, which many have set incorrectly. Neovim abides by it, and the debate is whether neovim should also ignore it like everyone else. See @tarruda 's longer reply: https://github.com/neovim/neovim/issues/2048#issuecomment-78546108

**Comment by chevex** [comment]
Ok that makes more sense. As much as I think it sucks that terminfo is set incorrectly, I guess count my vote for doing what everyone else is doing so users don't have to do this workaround.

**Comment by agauniyal** [comment]
@davidsteinberger yes and it dosen't works. This just changes my cursor from fat guy to slim guy but fails to delete any character.

**Comment by thomasboyt** [comment]
@badeip Following your advice for iTerm 2's hotkeys, I'm now able to use `C-h` in Neovim inside or outside of tmux. However, sending `C-h` to a terminal window in or outside of tmux just outputs `4;5u`, preventing me from navigating between tmux panes with `C-h`. Is there a terminal setting I'm missing that would fix this?

**Comment by badeip** [comment]
@thomasboyt it's not really my advice, I just described how to adapt @dragn suggestion to iTerm 2.
That said, it seems `libtermkey` has taken the pedantic approach to `c-h` et al.
That is probably "correct", but incompatible with several other programs such as `tmux` and `vim`.

IMO having to change terminal settings is not the right solution, as nvim should "just work".

A possible solution is to add a `libtermkey` run time setting with a "legacy mode" that can be switched off, and exposing this setting in nvim with e.g  `:set pedanticterm`
I have not looked at the code, so I do not know how feasible that is.

**Comment by justinmk** [comment]
Last several comments are incorrect. This is not a matter of libtermkey or neovim being pedantic. In fact, libtermkey will ignore terminfo if a terminal setting is detected, even if terminfo does not specify it.

In the case of ctrl-h specifically, there is a well-known quirk that neovim can possibly assume, to avoid this problem. 

I would request that further comments in this thread do not contain speculation. This thread is long enough.

**Comment by chevex** [comment]
Agreed ^

For those coming in, @tarruda's fix way above is the only "complete" workaround as it directly fixes the kbs quirk in terminfo. It's that quirk that most other programs referenced (including classic vim) are choosing to ignore, which is what neovim may ultimately end up doing.

**Comment by badeip** [comment]
@justinmk thanks for clearing that up, the issue is not very comprehensible for the uninitiated.
This thread contains several suggestions of how to work around this issue, but AFAIK no solution (including @tarruda's fix) that works for all and doesn't break e.g compatibility with `vim`

If you do not want more comments in this thread, I suggest either documenting the issue and workaround on the wiki and/or `:help c-h`

**Comment by justinmk** [comment]
>  I will modify neovim TUI to preprocess DEL and ctrl+h before delivering these bytes to libtermkey.

@tarruda Can you do this for 0.1? I don't think libtermkey is going to get the hooks needed any time soon.

**Comment by leonerd** [comment]
libtermkey isn't going to fix this one, as it's dead, EOL.

Some initial thoughts on a blueprint for what tickit might do in that direction can be found
  https://blueprints.launchpad.net/libtickit/+spec/ctrl-h-as-backspace

Comments welcome.

**Comment by gnachman** [comment]
@leonerd libtickit looks interesting. But it says it'll use libtermkey, so I'm confused about how dead libtermkey really is?

**Comment by leonerd** [comment]
@gnachman libtermkey can't continue being a standalone library. My plan is to copypaste the code across into tickit, so it can be expanded and given new abilities it currently can't have.

**Comment by smasher816** [comment]
I remember seeing something very similar when I was reading about suckless.org's st program. If anyone is interested in some of the backstory for why backspace sends ^H in terminfo I would encourage you to read this:

http://git.suckless.org/st/tree/FAQ#n101

**Comment by tarruda** [comment]
@smasher816 interesting read, thanks for the link

**Comment by justinmk** [comment]
For reference, `stty` doesn't work because we removed this logic:

https://github.com/vim/vim/blob/master/src/os_unix.c#L3493-L3504

**Comment by justinmk** [comment]
Pushed back to 0.3, blocked by libtermkey/libtikit

**Comment by leonerd** [comment]
For my reference, can you clarify what exactly you're blocked waiting for? What feature(s) would unblock you, were they to exist?

**Comment by justinmk** [comment]
@leonerd Here are the tickets:

- https://blueprints.launchpad.net/libtickit/+spec/override-terminfo
- https://blueprints.launchpad.net/libtickit/+spec/ctrl-h-as-backspace
- https://blueprints.launchpad.net/libtickit/+spec/termkey-merge

**Comment by leonerd** [comment]
I have a new approach. A program to guide users through interactively fixing their terminfo entry for their terminal.

A quick hack of a perl script currently here => http://fpaste.scsys.co.uk/499049

If this principle seems to be one that works, I'll have a go at writing it out as a standalone C program instead. Then we can refer users here to help them fix the 'backspace' entry, along with any other ones we find need fixing. This all seems much more the right solution, than trying to apply code in neovim (or any other program) to try to second-guess the values in terminfo. Far better to help the user set the known-correct ones.

**Comment by jwhitley** [comment]
@leonerd I understand and sympathize with the impetus behind your approach, but this still seems like something that needs a fully programmatic solution.  Guided hoop-jumping is still hoop-jumping.

**Comment by mayhewluke** [comment]
My terminfo entry has `kbs=\177` and neovim still sees `<C-h>` as `<BS>`. This stuff is all way beyond my paygrade but it would appear that having a correct terminfo entry is **not** the solution here after all. @tarruda seems to have alluded to this [earlier](https://github.com/neovim/neovim/issues/2048#issuecomment-78546108) as well:

> The problem is that every terminal seems to send ASCII DEL regardless of what terminfo says

So @leonerd's new approach isn't going to cut it, unfortunately. I guess someone's just going to have to bite the bullet and create some kind of override for the ASCII codes all these terminals are sending?

**Comment by leonerd** [comment]
@mayhewluke That sounds an interesting cornercase; one I wasn't expecting to observe.

Can you dig a little deeper on that? What does the libtermkey `demo` program report for that key?

**Comment by leonerd** [comment]
@mayhewluke - Hang on - before we continue, lets just clarify: Do you mean you are pressing the combination of keys "Ctrl" and "H", and neovim reports `<BS>`, or that you're pressing the key "Backspace" and it reports `<C-h>`? Your wording was a little ambiguous there.

**Comment by mayhewluke** [comment]
Sorry, let me clarify:

When I press `Ctrl` and `h` neovim reports `<BS>`. This is indicated by the following:

- Pressing `Ctrl+v` and then `Ctrl+h` in insert mode inserts `<BS>` into the document (as per [this comment](https://github.com/neovim/neovim/issues/2048#issuecomment-75637032))
- `nnoremap <BS> <C-w>h` is activated when pressing `Ctrl+h`

**OS:** NixOS 15.09
**Window Manager:** i3
**Terminal emulator:** Termite
**Shell:** zsh

```
$ echo $TERM
xterm-termite
$ infocmp $TERM | grep kbs
	kbs=\177, kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC,
$ tmux new-session
$ echo $TERM
screen-256color
$ infocmp $TERM | grep kbs
        il=\E[%p1%dL, il1=\E[L, ind=^J, initc@, is2=\E)0, kbs=\177,
```

> What does the libtermkey `demo` program report for that key?

I'm not sure what program you're referring to or how to run it, I'm afraid.

**Comment by leonerd** [comment]
@mayhewluke 

    wget http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.17.tar.gz
    tar -xf libtermkey-0.17.tar.gz
    cd libtermkey-0.17
    make
    ./demo

then press your Backspace key and see what it claims.
then press Ctrl-H and see what it claims.

**Comment by leonerd** [comment]
Oh. Hah.

This is a libtermkeybug. My bad... oops.

**Comment by leonerd** [comment]
... which I did fix at `-r353` but maybe neovim isn't using that, and is using release tarball?

**Comment by mayhewluke** [comment]
@leonerd NixOS 15.09 uses [libtermkey 0.17](http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.17.tar.gz) for neovim - not sure how to tell if that includes `-r353` or not though.

**Comment by leonerd** [comment]
If it's release 0.17 then no it won't be. I'm packing a 0.18 now though

**Comment by leonerd** [comment]
0.18 now up on http://www.leonerd.org.uk/code/libtermkey/

Also has big DEPRECATION warning but I know neovim already knows about that :)

**Comment by mayhewluke** [comment]
@leonerd I've built neovim using libtermkey 0.18 and it's working as expected based on the terminfo file. Thanks!

**Comment by leonerd** [comment]
Cool. :) I wonder if we can finally put 2048 to bed now?

Fun as it's been having such a memorable bug number though....

**Comment by mayhewluke** [comment]
@leonerd Although I'm glad to have a way of at least making this work, I do agree with @jwhitley - "Guided hoop-jumping is still hoop-jumping." Having to manually patch esoteric binary files through a copy+paste command isn't the sort of thing that gives me confidence in my toolchain or my understanding of how my setup works, and it's not a very user-friendly experience. Having just completely overhauled my OS and entire toolchain I can't even begin to count the number of hours I've spent googling for answers to arcane problems like this one. I'm not trying to point a finger at either libtermkey or neovim, but in order for neovim to have good user experience someone somewhere needs to take responsibility off of the user and solve it programmatically.

**Comment by leonerd** [comment]
@mayhewluke Which is why I built the 'fix-my-terminfo' tool.
It's a program you run, and it asks you a few *interactive* questions, guiding you to help it understand your terminal.

It asks you to press the backspace key, and captures the resulting bytes. It then prints a couple of terminal sequences that *should* result in the display being simply `123`, and asks the user to confirm if that is indeed the case.

Ultimately, that is the best and indeed perhaps only way to solve this. The latter parts are to fill in other missing pieces of terminfo string that are often missing. I can't honestly imagine any other way this could be done.

If you have a better suggestion to the problem of finding out what bytes the Backspace key would send, than asking the user to press it and capturing what bytes it *actually does* send, then do let me know.

**Comment by leonerd** [comment]
![fix-my-terminfo](https://cloud.githubusercontent.com/assets/350941/10338857/39690c70-6d00-11e5-9edb-0d0b825a9351.png)
^-- screenshot of it in action.

**Comment by fmoralesc** [comment]
@leonerd Is that tool available somewhere?

**Comment by leonerd** [comment]
@fmoralesc I'm intending it to be, yes. Currently it's a ~20 minute perl hack I wrote, so it has a few noncore dependencies so I could make it quickly. I'll make it more widely available once it's a bit easier for users to actually run it.

For now if you're curious, see

    $ bzr co http://bazaar.leonerd.org.uk/utils/fix-my-terminfo/
    $ cd fix-my-terminfo
    $ ./fix-my-terminfo

**Comment by ZyX-I** [comment]
@leonerd By the way, how about writing an utility which will configure various terminals to use fixterms standard for all or part of the key presses?

**Comment by joeynguyen** [comment]
Am I correct to assume that https://github.com/neovim/neovim/pull/3430 resolves this issue?  I upgraded my neovim just now, and renamed my 'xterm-256color.ti' (@tarruda's fix from March 10)  to 'xterm-256color.ti.bak', and opened a new Terminal window and opened neovim.  Ctrl-h seems to work fine for me now.

**Comment by leonerd** [comment]
@ZyX-I I'm not sure I see how this would work. Terminals vary massively in their methods of configuration; anywhere from custom files, Gnome's gconf, KDE's.. er.. whatever it has, Win32 registry,... hardware DIP switches, replacing EEPROM chips, maybe so hardcoded as to be impossible. No software tool running over that terminal could possibly hope to have any effect on that kind of configuration.

It could perhaps assist the user into verifying if their terminal was sending the right sequences, but there's nothing it could do to improve the situation if it was found not to be.

**Comment by ZyX-I** [comment]
@leonerd It is completely possible to script anything of this, except for hardware stuff and hardcoded. You just need an interface which takes from the user name of his terminal and sets of keys he wants to configure (maybe also the name of the profile: valid for konsole) and then either complain about unknown terminal or do changes to its configuration.

Note that most terminals use

1. xterm/\*rxvt entries in X resource database.
2. Just plain configuration files (it is “KDE’s.. er.. whatever”).
3. Gconf (VTE).
4. Whatever iTerm uses.

\: it may be enough to show some moves here, and first moves start with such tool. At least a few (maybe even one) enthusiast was enough to get true color support in many terminal emulators like vte, st and iTerm; and chances are better if you not just ask “lets have this feature”, but “I wrote a config, can you adopt it upstream to have this feature”. I also think that you have good chances of inclusion in KDE if you write configuration file for it, “Normal” one looks like

    keyboard "Normal"
    key Space+Ctrl : ""
    key Esc : "\E"
    key Tab+Shift-Ansi : "\t"
    key Tab+Shift+Ansi : "\E[Z"
    key Tab-Shift : "\t"
    key Backtab-Ansi : "\t"
    key Backtab+Ansi : "\E[Z"
    key Backspace : "\b"
    key Return+Shift : "\EOM"
    key Return-Shift+NewLine : "\r\n"
    key Return-Shift-NewLine : "\r"
    key Enter-NewLine : "\r"
    key Enter+NewLine : "\r\n"
    key Ins+AnyModifier : "\E[2;*~"
    key Ins-AnyModifier : "\E[2~"
    key Del+AnyModifier : "\E[3;*~"
    key Del-AnyModifier : "\E[3~"
    key Home+AnyModifier : "\E[1;*H"
    key Home+AppCursorKeys-AnyModifier : "\EOH"
    key Home-AppCursorKeys-AnyModifier : "\E[H"
    key End+AnyModifier : "\E[1;*F"
    key End+AppCursorKeys-AnyModifier : "\EOF"
    key End-AppCursorKeys-AnyModifier : "\E[F"
    key Left-Shift+Ansi+AnyModifier : "\E[1;*D"
    key Left-Shift+Ansi-AppCursorKeys-AnyModifier : "\E[D"
    key Left-Shift+Ansi+AppCursorKeys-AnyModifier : "\EOD"
    key Left-Shift-Ansi : "\ED"
    key Up+Shift-AppScreen : ScrollLineUp
    key Up-Shift+Ansi+AnyModifier : "\E[1;*A"
    key Up-Shift+Ansi-AppCursorKeys-AnyModifier : "\E[A"
    key Up-Shift+Ansi+AppCursorKeys-AnyModifier : "\EOA"
    key Up-Shift-Ansi : "\EA"
    key Right-Shift+Ansi+AnyModifier : "\E[1;*C"
    key Right-Shift+Ansi-AppCursorKeys-AnyModifier : "\E[C"
    key Right-Shift+Ansi+AppCursorKeys-AnyModifier : "\EOC"
    key Right-Shift-Ansi : "\EC"
    key Down+Shift-AppScreen : ScrollLineDown
    key Down-Shift+Ansi+AnyModifier : "\E[1;*B"
    key Down-Shift+Ansi-AppCursorKeys-AnyModifier : "\E[B"
    key Down-Shift+Ansi+AppCursorKeys-AnyModifier : "\EOB"
    key Down-Shift-Ansi : "\EB"
    key PgUp+Shift-AppScreen : ScrollPageUp
    key PgUp-Shift+AnyModifier : "\E[5;*~"
    key PgUp-Shift-AnyModifier : "\E[5~"
    key PgDown+Shift-AppScreen : ScrollPageDown
    key PgDown-Shift+AnyModifier : "\E[6;*~"
    key PgDown-Shift-AnyModifier : "\E[6~"
    key ScrollLock : ""
    key F1+AnyModifier : "\E[11;*~"
    key F1-AnyModifier : "\E[11~"
    key F2+AnyModifier : "\E[12;*~"
    key F2-AnyModifier : "\E[12~"
    key F3+AnyModifier : "\E[13;*~"
    key F3-AnyModifier : "\E[13~"
    key F4+AnyModifier : "\E[14;*~"
    key F4-AnyModifier : "\E[14~"
    key F5+AnyModifier : "\E[15;*~"
    key F5-AnyModifier : "\E[15~"
    key F6+AnyModifier : "\E[17;*~"
    key F6-AnyModifier : "\E[17~"
    key F7+AnyModifier : "\E[18;*~"
    key F7-AnyModifier : "\E[18~"
    key F8+AnyModifier : "\E[19;*~"
    key F8-AnyModifier : "\E[19~"
    key F9+AnyModifier : "\E[20;*~"
    key F9-AnyModifier : "\E[20~"
    key F10+AnyModifier : "\E[21;*~"
    key F10-AnyModifier : "\E[21~"
    key F11+AnyModifier : "\E[23;*~"
    key F11-AnyModifier : "\E[23~"
    key F12+AnyModifier : "\E[24;*~"
    key F12-AnyModifier : "\E[24~"

and is found in my system in `~/.kde4/share/apps/konsole/Normal.keytab` (to actually use it you need to replace existing/add `KeyBindings=Normal` to target `*.profile` file in the same location). KDE has compatibility layouts for XFree 4 and linux and solarized kernel terminals.

**Comment by expipiplus1** [comment]
@mayhewluke @leonerd I've submitted a PR to nixpkgs NixOS/nixpkgs#10393 to use version 0.18 of libtermkey.

As an aside, I still get `<BS>` when pressing `C-v C-h` in insert mode...

**Comment by sondr3** [comment]
Any updates on this? I still have this issue and it's incredibly frustrating.

**Comment by justinmk** [comment]
@tarruda has an ambitious project in mind that may avoid this issue entirely. He is working on it for 0.2.

**Comment by sondr3** [comment]
Cool, it's a bit annoying having to hack around the issue with various workarounds. Can't wait :smile:

**Comment by rainerborene** [comment]
@sondr3 Until then just follow @tarruda [suggestion](https://github.com/neovim/neovim/issues/2048#issuecomment-78045837).

**Comment by originalsouth** [comment]
@justinmk now I'm really curious what @tarruda's ambitious project is :)

**Comment by leonerd** [comment]
@sondr3 - if you're using latest neovim and your terminal is correct and your terminfo file correctly describes its behaviour, then it should work.

If it still does not, then at least one of these three facts must be false. So go fix those.

**Comment by leonerd** [comment]
@justinmk I do hope @tarruda's idea isn't just a reinvention of my `fix-my-terminfo` script. That already works to solve all this for all programs ever.

**Comment by justinmk** [comment]
@leonerd  we need a way to override terminfo codes. Even if distros fix their terminfo, this doesn't help users of older distros.

**Comment by leonerd** [comment]
@justinmk Exactly. A *USER* runs my program, presses the backspace key when the program asks them to, and then that program writes a customised terminfo file in the user's $HOME directory. That's it. That's all they need to do and now it's fixed for every program they'll ever run.

They run  `fix-my-terminfo`, and they press backspace when asked to.

What more could be simpler?

**Comment by tarruda** [comment]
> @justinmk I do hope @tarruda's idea isn't just a reinvention of my fix-my-terminfo script. That already works to solve all this for all programs ever.

@leonerd I have great respect for your work. As I said before, I think libtickit and libtermkey are amazing projects with great code quality, but this "ambitious project" they are talking about is basically a rework of the TUI that parses input without libtermkey, using unibilium for obtaining initial terminfo data but allowing users have full control over the strings.

I wanted to avoid implementing this, especially since libtermkey already does the hard work, but decided to do it anyway for the following reasons:

- As stated above, we want to let users override terminfo information about keys. As we all know, terminfo is broken on some distros, and requiring users to run an interactive script on every machine they log into is cumbersome compared to setting some options in the vimrc(which are more easily propagated when kept in a dot files repo).
- You already stated on #tickit that you have no plans add override mechanisms to libtermkey
- libtermkey will be merged into libtickit, and ATM we don't need any libtickit widget functionality(Neovim already contains logic to draw on each cell)
- The new TUI version will be mostly written in lua(some parts will still be C, especially the screen state manipulations and unibilium bindings). Lua is very high level, and with its builtin pattern matching library, parsing terminal codes will be trivial.

**Comment by rdlugosz** [comment]
> What more could be simpler?

While I agree that the _correct_ thing to do is to have users fix their terminfo files, the _simpler_ thing is to have neovim do what vim and other programs do & fix this internally. There may be a technical reason why neovim cannot do this, but if not I think following the lead of other applications is completely reasonable.

*Is there a reason why doing what classic vim does is not an option?* I re-read the thread before typing this question, but please excuse me if I missed it.

**Comment by bgourlie** [comment]
>What more could be simpler?

I also want to point out this is simple once the user:

- Realizes something isn't behaving as expected
- Googles for a solution to the problem
- Happens to discover this script and presumably determines it's the best solution to fixing the problem, among the various other suggested fixes

**Comment by tarruda** [comment]
> Is there a reason why doing what classic vim does is not an option? I re-read the thread before typing this question, but please excuse me if I missed it.

The reason those options were dropped in the first TUI refactor is because a major goal of Neovim is to decouple the UI from the core, which had those options tightly coupled.

My next PR will give back the user full control over how nvim interacts with the terminal, but it will be done with plain `g:` variables, which after #3603 will have the same power as the builtin options(plugins will be able to attach logic to change events). This will make even more sense because the TUI will be just another plugin from Neovim POV.

**Comment by tarruda** [comment]
BTW, even though libtermkey will no longer be used, Neovim will still support its key encoding scheme.

**Comment by rdlugosz** [comment]
Thx. I figured there was a reason for the change. So once libtermkey is no longer being used will this issue simply go away? If so, great! If not, I'd suggest that it's reasonable—desirable, even—to have internal handling of this issue at a layer that federates between the UI, the core & the terminal. (excuse my lack of knowledge of the Neovim architecture... speaking generally here & assuming you'll understand what I mean.)

**Comment by leonerd** [comment]
@rdlugosz (neo)vim can't really do what "most other programs do", because most other programs don't care to distinguish Backspace from Ctrl-H. I went off and researched a great many of them and all of the likely candidates (mutt, irssi, screen, tmux,...) conflate the two. vim is almost unique in all the world in offering any consideration of the fact that Backspace and Ctrl-H might actually be different.

What legacy vim does is to ask termios what its opinion is of the Backspace key. In my opinion that's only trading one possibly-wrong database (terminfo) for another (termios). I took the view that the only 100% correct foolproof way to know what byte the Backspace key sends is to ask the user to actually press the Backspace key and record what byte it actually sent.

**Comment by leonerd** [comment]
@bgourlie Do you have a different suggestion then? I'm very keen to hear it - I only came up with the current terminfo-fixing script because of a total lack of any better ideas after months of milling it around here. It's possible I haven't heard your suggestion though.

**Comment by rdlugosz** [comment]
Makes sense. Perhaps this is something neovim can do on startup and tag in `viminfo`?

Want to be clear that I appreciate that you've done a lot of research & work on this! My $0.02 is being kicked in to make sure we're keeping things as simple as possible for the user. I & others have found this thread went through the process described [above](https://github.com/neovim/neovim/issues/2048#issuecomment-154455956), which isn't a great user experience.

Totally understand that the idea of Neovim is to shed much of the old baggage & improve the architecture, but I think for the foreseeable future people will be looking for things that don't quite work right. When they stumble across something like their `C-h` mapping not working they're either going to research & find this solution or (most likely) they'll decide "meh, neovim is still a little buggy; needs more time" and go back to vim. Best to minimize the chances of that sort of thing happening.

**Comment by leonerd** [comment]
@rdlugosz How about a "first run" setup wizard? The first time a user runs neovim on a new $TERM value, it says:

    Hey, so this is the first time you've run neovim on this kind of terminal.
    Before we start, lets just check everything is working.
    Please press the <Backspace> key now, so I can check it's all fine

**Comment by badeip** [comment]
What about adding a new argument to nvim, such as `nvim --configure`?
It's not very vim-ish, but at least we don't have to bother all new users with a setting they might not care about.

**Comment by bgourlie** [comment]
@leonerd I'm not saying there's a better solution, I'm just pointing out it's only simple once the user realizes it's the proper solution.  I do like the idea of a first-run wizard.

**Comment by leonerd** [comment]
@bgourlie But that's how I phrased my question. ;) I know it's not absolutely simple - but it's the simplest thing I could practically create that actually works. I'm not aware of a simpler solution.

**Comment by bgourlie** [comment]
@leonerd There are really two issues -- The technical solution, which you've solved.  The other issue is the usability issue, and I think that's what most of this discussion revolves around.  What is best for the users?  Should we expect them to run a script so that they have the ideal *technical* solution, or a more pragmatic solution that will "just work" for most people but may not be technically ideal.

**Comment by rdlugosz** [comment]
I think a first run (or new terminal) wizard is a very reasonable idea.

**Comment by leonerd** [comment]
@tarruda:

>    As stated above, we want to let users override terminfo information about keys. As we all know, terminfo is broken on some distros, and requiring users to run an interactive script on every machine they log into is cumbersome compared to setting some options in the vimrc(which are more easily propagated when kept in a dot files repo).
>    You already stated on #tickit that you have no plans add override mechanisms to libtermkey
    libtermkey will be merged into libtickit, and ATM we don't need any libtickit widget functionality(Neovim already contains logic to draw on each cell)

I said I have no immediate plans to do that right now because I am busy doing other things and I don't *personally* need it. However, there's already an LP blueprint for considering this:

  https://blueprints.launchpad.net/libtickit/+spec/override-terminfo

You're welcome to comment on that, or take it and implement it yourself, or whatever. I said this isn't a priority for *me*, not that I'll block anyone else doing it.

I would hate to see neovim get annoyed and drop using this simply because of some small upset about Ctrl-H or whatever related reason it was.

>    The new TUI version will be mostly written in lua(some parts will still be C, especially the screen state manipulations and unibilium bindings). Lua is very high level, and with its builtin pattern matching library, parsing terminal codes will be trivial.

Well, of course my reply to that would be that it is (surely) fairly easy to call into a C library from Lua.

Having implemented all of this once, I'm aware how nontrivial parts of it can be. I really would like us to end up in a situation where neovim was using libtickit - not least because that reduces the total amount of work. But moreover, that leaves neovim able to much easier make use of more things that get added into libtickit as time goes on.

It's no great secret that I would *like* more terminals using `libvterm` and more application using `libtickit`. The more that do, the more will interoperate nicely and make use of yet newer abilities I start adding in.

**Comment by tarruda** [comment]
> I would hate to see neovim get annoyed and drop using this simply because of some small upset about Ctrl-H or whatever related reason it was.

@leonerd I did not mean to put any pressure for you to implement these features just for the sake of neovim. In the future libtickit widgets may be useful, but ATM the terminal UI does not know anything about widgets, it simply receives direct cell by cell drawing instructions. This means there's basically no difference from using libtickit to using unibilium directly, we would just be adding a unnecessary layer.

The only part of libtickit that is currently useful for neovim is the input layer which used to be implemented in libtermkey(and now you moved or will move into libtickit), but even so I'm forced to add hacks in order to work around some of its limitations.

One example of this(beyond not allowing terminfo override) is the fact that the only way it will stop trying to recognize `\x1b` as the alt modifier is to enable 8-bit mode(which is not an option for many terminal emulators). This has caused some users trouble with tmux mappings that  send `\x1b` expecting to leave insert mode(they began seeings strange characters printed to the screen since tmux sends keys with almost no delay). So I added  workaround code that handles `\x1b` specially before calling libtermkey for:

- Allowing `\x1b` followed by `\x00` to be interpreted as the esc key(this allowed tmux users to adapt their mappings for neovim)
- Parsing bracketed paste sequences which AFAIK is also not supported by libtermkey.

Implementing key parsing in lua will greatly clean up the current code.

> You're welcome to comment on that, or take it and implement it yourself, or whatever. I said this isn't a priority for me, not that I'll block anyone else doing it.

I also prefer to call C code from lua, but implementing the changes I require into libtickit would require a lot more than what I have planned. Here's what comes to mind(besides adding/debugging new C code, which is slower than doing in lua):

- Get familiar with libtickit architecture and code patterns. I doubt you will want quick and dirty patches that simply implement the feature, it needs to follow your standards(BTW this is not a criticism)
- Maintain a fork of libtickit until the changes are reviewed and merged upstream. This will require package maintainers to target the fork.
- Once it is merged, switch back to using upstream, which is also more work for package maintainers.

As I said before, this doesn't mean we won't ever use libtickit again, it just means ATM it is not the best choice. If eventually it ends up meeting our requirements while providing a maintenance improvement, I don't see why we couldn't switch to using it.

**Comment by mcmire** [comment]
+1 for [this comment](https://github.com/neovim/neovim/issues/2048#issuecomment-78045837) as a workaround for this issue.

**Comment by selurvedu** [comment]
I confirm the following workaround:
1) libtermkey 0.18 (doesn't work with 0.17).
2) [Patched](https://github.com/neovim/neovim/issues/2048#issuecomment-78045837) `~/.terminfo`.

@tarruda, @leonerd – thank you, guys!

**Comment by sublimecoder** [comment]
The issue still exists for me, currently using iterm2 escape sequence key mapping as a workaround.

**Build information**
NVIM v0.1.0-157-g0ee3398 (compiled Dec  3 2015 16:14:20)
Commit: 0ee33981e8967bab75427214b432e22ebcf40d6f

**Comment by leonerd** [comment]
@sublimecoder can you explain "iterm2 escape sequence key mapping"?

**Comment by sublimecoder** [comment]
@leonerd Here are the steps to add the key mapping. 

Edit -> Preferences -> Keys
Press +
Press Ctrl+h as Keyboard Shortcut
Choose Send Escape Sequence as Action
Type [104;5u for Esc+

referenced from @badeip's comment

**Comment by VorT3x** [comment]
@sublimecoder thanks a lot! Fixed my problem

**Comment by nabn** [comment]
When I follow the steps mentioned by @sublimecoder, the mapping works fine for most cases. But if I have a server running (webpack-dev-server, in this case), when I hit C-h, it prints out the mapped characters to the console, instead of navigating to another split in tmux.

**Comment by roychoo** [comment]
yeah i have the problem as @nabn , anyone found any workaround on this?

**Comment by jedahan** [comment]
export TERMINFO="$HOME/.terminfo" fixed it for me

**Comment by ahan** [comment]
@badeip thanks, fixed my problem at iTerm2 for ctrl+h, and need to mention that to reboot the iTerm2 to make it work.

**Comment by alesandar** [comment]
For anyone who uses st (by suckless), it seems like it has the proper terminfo already.
Appending `set -g default-terminal "st-256color"` to my configuration fixed everything.

**Comment by wsdjeg** [comment]
@alesandar I use st last week,but many hot key not works as expected.ctrl + home or ctrl + end do not works .so I have change to termite.everything works well

**Comment by aemb** [comment]
This is what I added to my `.vimrc` and it still doesn't work:
```
if has('nvim')
  " Hack to get C-h working in NeoVim
  nmap <BS> <C-W>h
endif
```
and  `verbose imap <BS>` gives me this result
```
i  <BS>        * lexima#expand('<BS>', 'i')
        Last set from ~/.vim/plugged/lexima.vim/autoload/lexima/insmode.vim
```

Any idea on how to get around this ? I'm not sure of what's happening there.

**Comment by fmoralesc** [comment]
@aemb You are mapping `<BS>` in normal mode, not insert mode, so the plugin takes over. It might override it even if you do map it in insert mode, but that will depend on how well behaved it is.

**Comment by blueyed** [comment]
For me this is only a problem with `sudo nvim`?!
With Neovim in tmux `:send-keys c-h` works as expected, but with `sudo nvim` the `:send-keys c-h` from tmux will become `<BS>` in Neovim.

**Comment by aemb** [comment]
@fmoralesc I used `imap` and it doesn't work.
```{vim}
if has('nvim')
  " Hack to get C-h working in NeoVim
  imap <BS> <C-W>h
endif
```

The result of `verbose imap <BS>` is the one expected this time but it still behaves bad.
`<C-h>` still does `<BS>`.
```{vim}
i  <BS>          <C-W>h
        Last set from ~/.vim/init.vim
```

**Comment by justinmk** [comment]
@aemb You must also `inoremap <C-W>h <something>`.

Please open troubleshooting questions in new issues or in other forums. This issue does not need more discussion about individual configs.

**Comment by trusktr** [comment]
@aemb See `:help map`, then learn which mode you want to map for (n = normal, i = insert, etc) and what command you want to map `<bs>` to in that mode.

**Comment by siphiuel** [comment]
I've started having this issue in Tmux+Neovim after i've modified

    set -g default-terminal "xterm-256color"

to 

    set -g default-terminal "screen-256color"

in `.tmux.conf`.

Kudos to @jedahan - `export TERMINFO="$HOME/.terminfo"` fixed it for me as well.

**Comment by blueyed** [comment]
FWIW, I've found this in tmux source:

```c
/*
 * Check for backspace key using termios VERASE - the terminfo
 * kbs entry is extremely unreliable, so cannot be safely
 * used. termios should have a better idea.
 */
bspace = tty->tio.c_cc[VERASE];
if (bspace != _POSIX_VDISABLE && (key & KEYC_MASK_KEY) == bspace)
	key = (key & KEYC_MASK_MOD) | KEYC_BSPACE;
```
https://github.com/tmux/tmux/blob/fe4e9470bb504357d073320f5d305b22663ee3fd/tty-keys.c#L625-L632

And it gets set for panes initially: `tio2.c_cc[VERASE] = '\177';` (https://github.com/tmux/tmux/blob/fe4e9470bb504357d073320f5d305b22663ee3fd/window.c#L866).

Not sure if it's really helpful, but might give some idea to fix this.

**Comment by justinmk** [comment]
@blueyed Yes, that's the traditional way to avoid this issue (Vim uses the same approach): fall back to termios. We would like to do that, but it requires forking libtermkey (or more ideally, getting a hook in libtickit when libtermkey is merged into libtickit).

**Comment by dojoteef** [comment]
I know I'm late to the party, but if anyone is looking for a solution with a more limited scope the approach I used will gracefully fallback to termios settings if needed for your terminal without requiring a global setting.

https://github.com/dojoteef/dotfiles/blob/master/init/50_nvim.sh#L39
https://github.com/dojoteef/dotfiles/blob/master/source/50_editor.sh#L4

**Comment by justinmk** [comment]
For those finding this issue, the fix is covered in the FAQ:

https://github.com/neovim/neovim/wiki/FAQ#my-ctrl-h-mapping-doesnt-work

**Comment by justinmk** [comment]
If anyone having this issue can build https://github.com/neovim/neovim/pull/5758 and test it, that would be greatly appreciated.

**Comment by justinmk** [comment]
This is fixed on master (not released yet). To build from source, you probably need to:

    rm -rf .deps build

However master is unstable currently, so most users should wait until the next release.
